Implemented re-resolve of endpoints when android network changes

spacecowboy opened the pull request:
This commit adds a re-resolving of peer endpoints when networks, such
as WiFi or Cellular, comes or goes.

I have tested the userspace implementation on my own phone but I don't
have access to a device which contains the kernel module so the
WgQuick-version is so far untested.

I also sent it as a patch to the mailing list if that is preferred but it's awaiting
moderation approval.

msfjarvis commented:
GitHub is fine for wireguard-android changes as I'm typically the one reviewing them. This looks okay on a cursory glance, but you will need to amend the commit to add a Signed-off-by line to the message with your real name (running `git commit --amend --signoff` and force pushing this branch should be enough) before we can accept it.

msfjarvis commented on tunnel/src/main/java/com/wireguard/android/backend/GoBackend.java:
```diff
@@ -202,6 +204,27 @@ public State setState(final Tunnel tunnel, State state, @Nullable final Config c
+            return;
+        }
+
+        final String tunnelName = tunnel.getName();
+        final Config config = currentConfig;
+        if (config == null) {
+            return;
+        }
+
+        reResolveEndpointsInternal(tunnelName, config);
```
```suggestion
        final Config config = currentConfig;
        if (config == null) {
            return;
        }

        reResolveEndpointsInternal(tunnel.getName(), config);
```

msfjarvis commented on tunnel/src/main/java/com/wireguard/android/backend/WgQuickBackend.java:
```diff
@@ -162,6 +162,23 @@ public State setState(final Tunnel tunnel, State state, @Nullable final Config c
+            return;
+        }
+
+        final String tunnelName = tunnel.getName();
+        final Config config = runningConfigs.get(tunnel);
+        if (config == null) {
+            return;
+        }
+
+        reResolveEndpointsInternal(tunnelName, config);
```
```suggestion
        final Config config = runningConfigs.get(tunnel);
        if (config == null) {
            return;
        }

        reResolveEndpointsInternal(tunnel.getName(), config);
```

msfjarvis commented on tunnel/src/main/java/com/wireguard/config/Config.java:
```diff
@@ -182,6 +182,51 @@ public String toWgUserspaceString() {
+                sb.append(' ').append(peerString);
+            }
+        }
+        if (gotPeers) {
+            return sb.toString();
+        }
```
```suggestion

        if (gotPeers) {
            return sb.toString();
        }
```

msfjarvis commented on tunnel/tools/libwg-go/api-android.go:
```diff
@@ -200,4 +200,29 @@ func wgVersion() *C.char {
+		return -1
+	}
+
+	logger.Info.Println("Re-resolved peer endpoints")
```
```suggestion
	logger.Info.Println("Re-resolved peer endpoints successfully")
```

msfjarvis commented on tunnel/tools/libwg-go/jni.c:
```diff
@@ -69,3 +70,25 @@ JNIEXPORT jstring JNICALL Java_com_wireguard_android_backend_GoBackend_wgVersion
 	return ret;
 }
+
+JNIEXPORT jint JNICALL Java_com_wireguard_android_backend_GoBackend_wgSet(JNIEnv *env, jclass c, jstring ifname, jint handle, jstring settings)
```
I'd expect this to log on success as well if the Go backend version does, but perhaps its pointless with the logging RootShell already does.

spacecowboy replied:
Yeah maybe. Let's wait for Jason's comments on that section before deciding

msfjarvis commented on ui/src/main/java/com/wireguard/android/Application.kt:
```diff
@@ -94,6 +99,41 @@ class Application : android.app.Application() {
+    private fun registerForNetworkChanges() {
+        val cm = applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+        val callback = object : ConnectivityManager.NetworkCallback() {
```
should we be unregistering the callback in `onTerminate`?

spacecowboy replied:
It should make no functional difference as the listener is tied to the application lifetime, but I can see the value of adding it anyway

msfjarvis reviewed (commented):
A handful of pedantic comments, the implementation looks reasonable on the Android userspace end of things. Jason will review the Go and C bits when he's available.

spacecowboy commented:
Thanks for the review @msfjarvis . I have addressed the comments, added the sign-off, and added a couple of tests for the config string generation functions.

msfjarvis commented:
> Thanks for the review @msfjarvis . I have addressed the comments, added the sign-off, and added a couple of tests for the config string generation functions.

Great. I talked a bit with @zx2c4 about this and he's of the opinion that this potentially introduces a security hole where an attacker can briefly disrupt a user's WiFi connection then poison the resolver cache. When the device reconnects, it will resolve DNS once more and now point to an attacker controlled machine. If you could explain in some detail as to what precise problem you're facing and aim to resolve with this PR we can work towards a safer solution. Jason also noted that the method naming appears to be inherited from the iOS codebase, where the reason to do this is to be able to resolve XLAT information that isn't necessary on Android due to the clatd daemon doing the correct thing transparently.

spacecowboy commented:
> Great. I talked a bit with @zx2c4 about this and he's of the opinion that this potentially introduces a security hole where an attacker can briefly disrupt a user's WiFi connection then poison the resolver cache. When the device reconnects, it will resolve DNS once more and now point to an attacker controlled machine. If you could explain in some detail as to what precise problem you're facing and aim to resolve with this PR we can work towards a safer solution. Jason also noted that the method naming appears to be inherited from the iOS codebase, where the reason to do this is to be able to resolve XLAT information that isn't necessary on Android due to the clatd daemon doing the correct thing transparently.

Sure no problem. This is my attempt at implementing the existing [reresolve-dns script](https://git.zx2c4.com/wireguard-tools/tree/contrib/reresolve-dns) to solve the issue that a server might resolve to different ip-addresses depending on if you are inside or outside a local network (e.g. WiFi).

While I have configured my router to always resolve my hostname to the external ip, and to use resolution mirroring (or whatever it's called) I have still experienced issues with wireguard for android whenever my wifi connectivity changes. In short, the moment I lose or gain Wif, depending on the state when I initially started the tunnel, I will lose all internet connectivity on my phone since wireguard will fail to re-resolve the dns name. It's particularly annoying since my living room is at the edge of one of the wifi networks.

Others around the internet have also requested this functionality, for example here: https://www.reddit.com/r/WireGuard/comments/evkspf/reresolvedns_for_android/

The security issue might be real, my suggestion would be to add a setting for this feature in that case. With a description recommending against enabling it unless it is actually necessary.

zx2c4 commented:
Thanks for this contribution. However, I'm not sure it totally makes sense and I have a few questions about its intent and design:

- Why reresolve based on network changes? What does that have to do with DNS? Outside of XLAT (which isn't an issue on Android thanks to clatd, though it is on iOS) what does this accomplish? Why are network change notifications the trigger?
- The lifetime of a DNS entry is usually its TTL, right?
- Won't the reresolutions happen inside the tunnel if 0.0.0.0/0 is routed, in which case, if it works, there was no reason for reresolution in the first place.
- But in resolving that prior point, I have doubts about the security of making resolutions outside the tunnel while the tunnel is up.

It's one thing to release a small script people can run on their Linux boxes to add additional finicky functionality, but it's quite another to add this as a first class feature into the client programs, and an even bigger deal to introduce a configuration knob, as you've suggested.

spacecowboy commented:
>     * Why reresolve based on network changes? What does that have to do with DNS? Outside of XLAT (which isn't an issue on Android thanks to clatd, though it is on iOS) what does this accomplish? Why are network change notifications the trigger?

No idea what XLAT is. As I explained it is the transition between WiFi and Cellular which creates the resolution problem so doing the re-resolution based on network changes is spot on.

A DNS name can resolve differently depending on the network used, in case the wireguard peer is inside the local network when on WiFi. So example

on wifi: `wg-server.test.com` resolves to `192.168.1.1`
on cellular: `wg-server.test.com` resolves to `200.100.99.32`
 
>     * The lifetime of a DNS entry is usually its TTL, right?

Sure but that is irrelevant when different DNS servers respond differently

>     * Won't the reresolutions happen inside the tunnel if 0.0.0.0/0 is routed, in which case, if it works, there was no reason for reresolution in the first place.

No it is the peer endpoint which needs to be updated. The tunnel is non-functional until the endpoint is updated, and the device loses all internet connectivity until that happens.

The resolution is done when the config is parsed and transformed to userspace / wg-quick formats here: https://github.com/WireGuard/wireguard-android/blob/56a4862442e5295d76d886523c87eeabcf74288d/tunnel/src/main/java/com/wireguard/config/InetEndpoint.java#L90

and once the tunnel is up, the endpoint is never resolved again.
 
>     * But in resolving that prior point, I have doubts about the security of making resolutions outside the tunnel while the tunnel is up.

It might be that the security scenario you describe is vital to defend against - I consider it "above my paygrade" as it were.

But in this scenario total security is achieved since the tunnel is dead until it is manually restarted, or wifi is restored / disabled. When the user is forced to manually restart the tunnel then the highjacked DNS scenario is still possible.

> It's one thing to release a small script people can run on their Linux boxes to add additional finicky functionality, but it's quite another to add this as a first class feature into the client programs, and an even bigger deal to introduce a configuration knob, as you've suggested.

We might have differing opinions about how big a deal it is, but the fact remains that you can't run a small script on android to achieve this.

zx2c4 commented:
> the fact remains that you can't run a small script on android to achieve this.

You actually can.

> No it is the peer endpoint which needs to be updated. The tunnel is non-functional until the endpoint is updated, and the device loses all internet connectivity until that happens.

Do you mean "yes"? Asking again: the resolution will happen outside of the tunnel, right? Otherwise, how would it resolve?

> on wifi: wg-server.test.com resolves to 192.168.1.1
> on cellular: wg-server.test.com resolves to 200.100.99.32

Ahh, I understand the problem you're up against now. Sounds like the cleaner solution here would be to just fix your local DNS resolver?

spacecowboy commented:
> > the fact remains that you can't run a small script on android to achieve this.
> 
> You actually can

Would this correspond to "Allow remote control apps" in the settings? If it's achievable that way I don't mind moving this functionality to a separate app.

> Do you mean "it happens inside the tunnel" or "it happens outside the tunnel"? Asking again: the resolution must happen outside of the tunnel, right? Otherwise, how would it resolve? But if the user has routed 0.0.0.0/0 through the tunnel, won't the system force that resolution to happen inside the tunnel?

Ah. Now I get it. Intuitively that makes sense. Not sure exactly what that corresponds to on Android, but I don't use "Block connections without VPN" setting. So I guess I don't route everything through the tunnel?

I use wireguard primarily as a way to reach services in my internal network (NFS for example).

My first approach restarted the tunnel on network changes but that felt heavy.

> > on wifi: wg-server.test.com resolves to 192.168.1.1
> > on cellular: wg-server.test.com resolves to 200.100.99.32
> 
> Ahh, I understand the problem you're up against now. Sounds like the cleaner solution here would be to just fix your local DNS resolver?

I fiddled plenty with it but still experienced issues. That was some time ago though and it's possible that I changed even more settings since then which made it work. Or Android does indeed resolve outside of the tunnel in my case.

filippz commented:
I also have the need for re-resolving IP address - in my case because both peers have dynamic IP addresses so "server" peer has DDNS entry that gets updated at least once in 24h. I don't have the problem that IP is different when resolution occurs on different networks because I use NAT Hairpin in "local" WiFi so it doesn't matter if public IP is used instead of local IP of the "server" peer.

On Win10 I've written a small PowerShell script that checks if tunnel is up and if so if the "server" is pingable - if not I use DNS servers configured on interfaces other than the one of WireGuard.
On Android I would need to use the similar logic and Automate or similar app - the problem I've encountered is that I can't find a way to tell if WireGuard tunnel is active so I don't bring it UP if it wasn't so in the first place. Any hints?

zx2c4 commented:
This PR does not accomplish what you're after.
